packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -0,0 +1,358 @@
+      throw new Error('Given state does not match url query state')
+
+    return urlAccessCode
+  }
+
+  /**
+   * Exchanges an access code for an access token. This function does **not** update the client's token.
+   * @throws {NotRegisteredException} When the client doesn't have it's registration information
+   * @param   {string} accessCode The access code contained in the redirection URL — sett `client.getAccessCodeFromURL()`
+   * @returns {Promise} A promise that resolves with an AccessToken object.
+   */
+  async fetchAccessToken(accessCode) {
+    if (!this.isRegistered()) throw new NotRegisteredException()
+
+    const data = new URLSearchParams({
+      grant_type: 'authorization_code',
+      code: accessCode,
+      client_id: this.oauthOptions.clientID,
+      client_secret: this.oauthOptions.clientSecret
+    })
+
+    const result = await this.fetch('POST', '/auth/access_token', data, {
+      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+    })
+
+    return new AccessToken(result)
+  }
+
+  /**
+   * Retrieves a new access token by refreshing the currently used token.
+   * @throws {NotRegisteredException} When the client doesn't have it's registration information
+   * @throws {Error} The client should already have an access token to use this function
+   * @returns {Promise} A promise that resolves with a new AccessToen object
+   */
+  async refreshToken() {
+    if (!this.isRegistered()) throw new NotRegisteredException()
+    if (!this.token) throw new Error('No token to refresh')
+
+    const data = new URLSearchParams({
+      grant_type: 'refresh_token',
+      refresh_token: this.token.refreshToken,
+      client_id: this.oauthOptions.clientID,
+      client_secret: this.oauthOptions.clientSecret
+    })
+
+    const result = await this.fetch('POST', '/auth/access_token', data, {
+      headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+    })
+
+    return new AccessToken({
+      refresh_token: this.token.refreshToken,
+      ...result
+    })
+  }
+
+  /**
+   * Updates the client's stored token
+   * @param {string} token = null The new token to use — can be a string, a json object or an AccessToken instance.
+   */
+  setCredentials(token = null) {
+    if (token) {
+      this.token = token instanceof AccessToken ? token : new AccessToken(token)
+    } else {
+      this.token = null
+    }
+  }
+
+  /**
+   * Turns the client's registration access token into a header suitable for HTTP requests. Used in some queries to manipulate the client on the server side.
+   * @returns {string}
+   * @private
+   */
+  registrationAccessTokenToAuthHeader() {
+    return 'Bearer ' + this.oauthOptions.registrationAccessToken
+  }
+}
+
+class NotRegisteredException extends Error {
+  constructor(message = 'Client not registered or missing OAuth informations') {
+    super(message)
+    this.message = message
+    this.name = 'NotRegisteredException'
+  }
+}
```

Comment:
I know there's some weird stuff with classes extending the Error class. See https://stackoverflow.com/questions/31089801/extending-error-in-javascript-with-es6-syntax-babel

Reply:
Interesting, although in that case everything seems to work fine. The error messages are in the snapshots for example :)